processor/processor.go

```diff
@@ -68,70 +78,113 @@ func ProcessConstants() {
 		printTrace(fmt.Sprintf("nanoseconds build extension to language: %d", makeTimestampNano()-startTime))
 	}
 
-	startTime = makeTimestampMilli()
-	for name, value := range database {
-		complexityTrie := &Trie{}
-		slCommentTrie := &Trie{}
-		mlCommentTrie := &Trie{}
-		stringTrie := &Trie{}
-		tokenTrie := &Trie{}
-
-		complexityMask := byte(0)
-		singleLineCommentMask := byte(0)
-		multiLineCommentMask := byte(0)
-		stringMask := byte(0)
-		processMask := byte(0)
-
-		for _, v := range value.ComplexityChecks {
-			complexityMask |= v[0]
-			complexityTrie.Insert(T_COMPLEXITY, []byte(v))
-			if !Complexity {
-				tokenTrie.Insert(T_COMPLEXITY, []byte(v))
-			}
-		}
-		if !Complexity {
-			processMask |= complexityMask
+	// If lazy is set then we want to load in the features as we find them not in one go
+	// however otherwise being used as a library so just load them all in
+	if !isLazy {
+		startTime = makeTimestampMilli()
+		for name, value := range languageDatabase {
+			processLanguageFeature(name, value)
 		}
 
-		for _, v := range value.LineComment {
-			singleLineCommentMask |= v[0]
-			slCommentTrie.Insert(T_SLCOMMENT, []byte(v))
-			tokenTrie.Insert(T_SLCOMMENT, []byte(v))
+		if Trace {
+			printTrace(fmt.Sprintf("milliseconds build language features: %d", makeTimestampMilli()-startTime))
 		}
-		processMask |= singleLineCommentMask
+	} else {
+		printTrace("configured to lazy load language features")
+	}
+}
 
-		for _, v := range value.MultiLine {
-			multiLineCommentMask |= v[0][0]
-			mlCommentTrie.InsertClose(T_MLCOMMENT, []byte(v[0]), []byte(v[1]))
-			tokenTrie.InsertClose(T_MLCOMMENT, []byte(v[0]), []byte(v[1]))
-		}
-		processMask |= multiLineCommentMask
+// Will load a single feature as requested given the name
+// this is used with lazy loading
+func LoadLanguageFeature(loadName string) {
+	if !isLazy {
+		return
+	}
 
-		for _, v := range value.Quotes {
-			stringMask |= v[0][0]
-			stringTrie.InsertClose(T_STRING, []byte(v[0]), []byte(v[1]))
-			tokenTrie.InsertClose(T_STRING, []byte(v[0]), []byte(v[1]))
-		}
-		processMask |= stringMask
-
-		LanguageFeatures[name] = LanguageFeature{
-			Complexity:            complexityTrie,
-			MultiLineComments:     mlCommentTrie,
-			SingleLineComments:    slCommentTrie,
-			Strings:               stringTrie,
-			Tokens:                tokenTrie,
-			Nested:                value.NestedMultiLine,
-			ComplexityCheckMask:   complexityMask,
-			MultiLineCommentMask:  multiLineCommentMask,
-			SingleLineCommentMask: singleLineCommentMask,
-			StringCheckMask:       stringMask,
-			ProcessMask:           processMask,
+	// Check if already loaded and if so return because we don't need to do it again
+	LanguageFeaturesMutex.Lock()
+	_, ok := LanguageFeatures[loadName]
```

Comment:
I wonder if there's any argument for using a [syncmap](https://godoc.org/golang.org/x/sync/syncmap) here. Not sure if there'd be enough contention on the lock to justify it, but you never know.

Reply:
I did look into that briefly at some point. At the time it was slower than just putting in a mutex. In this case I was being a little lazy though because it was easier to just add it in then change everything around it. 

At some point in the future it might be worth trying it out again as it would make the code simpler which I am always in favor of.

Reply:
I tried it quickly and got the same result of being very slightly slower.